pkg/offline/offline.go

```diff
@@ -77,40 +78,81 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
 	}, nil
 }
 
-// Sync returns a function to send queued heartbeats to the WakaTime API.
+// Sync returns a function to send all queued heartbeats to the WakaTime API.
 func Sync(filepath string, syncLimit int) func(next heartbeat.Handle) error {
 	return func(next heartbeat.Handle) error {
 		log.Debugf("execute offline sync with file %s", filepath)
 
-		hh, err := popHeartbeats(filepath, syncLimit)
-		if err != nil {
-			return fmt.Errorf("failed to fetch heartbeat from offline queue: %s", err)
-		}
+		var (
+			count   int
+			exclude []string
+		)
 
-		if len(hh) == 0 {
-			log.Debugln("abort execution, as there are no queued heartbeats ready for sending")
+		for {
+			if count > maxSyncRequests {
+				return fmt.Errorf("exceeded max sync request limit of %d", maxSyncRequests)
+			}
 
-			return nil
-		}
+			opts := PopOptions{
+				Excluded: exclude,
+				Limit:    syncLimit,
+			}
 
-		results, err := next(hh)
-		if err != nil {
-			requeueErr := pushHeartbeats(filepath, hh)
-			if requeueErr != nil {
-				log.Errorf("failed to push heatbeats after api error to queue: %s", requeueErr)
+			hh, err := popHeartbeats(filepath, opts)
+			if err != nil {
+				return fmt.Errorf("on pass #%d, failed to fetch heartbeat from offline queue: %s", count, err)
+			}
+
+			if len(hh) == 0 {
+				log.Debugf("on pass #%d, abort execution, as there are no queued heartbeats ready for sending", count)
+
+				break
 			}
 
-			return err
+			results, err := next(hh)
+			if err != nil {
+				requeueErr := pushHeartbeats(filepath, hh)
+				if requeueErr != nil {
+					log.Errorf("on pass #%d, failed to push heatbeats after api error to queue: %s", count, requeueErr)
+				}
+
+				return err
+			}
+
+			invalidResults, err := handleResults(filepath, results, hh)
+			if err != nil {
+				return fmt.Errorf("on pass #%d, failed to handle heatbeats api results: %s", count, err)
+			}
+
+			var ignored int
+
+			for _, i := range invalidResults {
+				// ignore "too many heartbeats" error results here, to send them again.
+				if i.Status == http.StatusTooManyRequests {
+					ignored++
+					continue
+				}
+
+				exclude = append(exclude, i.Heartbeat.ID())
```

Comment:
We only want to exclude heartbeats on `400` response code. Every other response code (`429`, `500`, etc.) we want to add the heartbeats back to the offline db.